crates/circuit/src/operations.rs

```diff
@@ -35,6 +36,53 @@ pub enum Param {
     Obj(PyObject),
 }
 
+impl Param {
+    pub fn add(&self, py: Python, other: &Param) -> Param {
+        match [self, other] {
+            [Self::Float(a), Self::Float(b)] => Param::Float(a + b),
+            [Self::Float(a), Self::ParameterExpression(b)] => Param::ParameterExpression(
+                b.clone_ref(py)
+                    .call_method1(py, intern!(py, "__radd__"), (*a,))
+                    .expect("Parameter expression addition failed"),
+            ),
+            [Self::ParameterExpression(a), Self::Float(b)] => Param::ParameterExpression(
+                a.clone_ref(py)
+                    .call_method1(py, intern!(py, "__add__"), (*b,))
+                    .expect("Parameter expression addition failed"),
+            ),
+            [Self::ParameterExpression(a), Self::ParameterExpression(b)] => {
+                Param::ParameterExpression(
+                    a.clone_ref(py)
+                        .call_method1(py, intern!(py, "__add__"), (b,))
+                        .expect("Parameter expression addition failed"),
+                )
+            }
+            _ => unreachable!(),
```

Comment:
Yeah I agree this definitely could occur depending on the caller. I think I added this for `global_phase` and we have validation for global phase that it's either `Param::ParameterExpresion` or `Param::Float` so this holds true for these, but since this is on `Param` itself we should make sure to handle other cases. What would you prefer for the case of other combinations of variants here? Return a `PyValueError`? Or should I try to call `__add__` with the arbitrary pyobject and return the resulting exception. I feel like I wanted to avoid the PyError for this for some reason but I can't remember what that was (probably to keep it simple for the pure float case and also because it's not a `pymethod`). The other option is I can remove this as a method and just have a helper function in `dag_circuit.rs` that does this.

Reply:
I think basically everything to do with `Param` at the moment can't escape Python space yet - we can try and pretend it's not there, but I think we're setting ourselves up for hard-to-debug problems in the future with symbolic parameters if we do.  (It's part of my motivation for wanting to split the case of "all parameters are floats" from "parameters are arbitrary `Param` instances in the instructions.)

I'm fine with either expanding this function to cover all cases and attempting to `a.add(b).extract::<Param>()` as the final step, or if it makes sense ot be more specialised for performance in one particular place, to inline this special case there.

Reply:
I made it a private function for working with global phase in: https://github.com/Qiskit/qiskit/pull/12550/commits/464e87b954ddb78e1faa82ed14aaeb1b6cb8e7b8